futures-util/src/try_future/try_join.rs

```diff
@@ -0,0 +1,131 @@
+#![allow(non_snake_case)]
+
+use crate::future::{MaybeDone, maybe_done};
+use crate::try_future::{TryFutureExt, IntoFuture};
+use core::fmt;
+use core::mem::PinMut;
+use futures_core::future::{Future, TryFuture};
+use futures_core::task::{self, Poll};
+
+macro_rules! generate {
+    ($(
+        $(#[$doc:meta])*
+        ($Join:ident, <Fut1, $($Fut:ident),*>),
+    )*) => ($(
+        $(#[$doc])*
+        #[must_use = "futures do nothing unless polled"]
+        pub struct $Join<Fut1: TryFuture, $($Fut: TryFuture),*> {
+            Fut1: MaybeDone<IntoFuture<Fut1>>,
+            $($Fut: MaybeDone<IntoFuture<$Fut>>,)*
+        }
+
+        impl<Fut1, $($Fut),*> fmt::Debug for $Join<Fut1, $($Fut),*>
```

Comment:
- It'd be nicer if the `Debug` impl was after the inherent impl

- Should there be an `Unpin` impl?

Reply:
Is there some reason `derive(Debug)` doesn't work for this struct, it looks like the same implementation and should be equivalent bounds?

Reply:
Done, also updated future/join.rs

I don't think we need an `Unpin` impl-- it should be the same as the inferred one from being an auto trait.

Reply:
These are not the same as the `derive` impls-- deriving fails as it doesn't know to put a bound on the associated `Ok` and `Error` types.

Reply:
`Unpin` is still an auto trait? I thought that changed o_O

Reply:
No.

Reply:
No as in "it isn't an auto trait anymore" or "it didn't change, it still is"?